// backendless/es/unit-of-work/op-result-value-reference.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OpResultValueReference = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var OpResultValueReference = /*#__PURE__*/function () {
  function OpResultValueReference(opResult, _ref) {
    var resultIndex = _ref.resultIndex,
        propName = _ref.propName;
    (0, _classCallCheck2["default"])(this, OpResultValueReference);
    this.opResult = opResult;
    this.resultIndex = resultIndex;
    this.propName = propName;
  }

  (0, _createClass2["default"])(OpResultValueReference, [{
    key: "getTableName",
    value: function getTableName() {
      return this.opResult.getTableName();
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      var result = this.opResult.toJSON();

      if (this.resultIndex !== undefined) {
        result.resultIndex = this.resultIndex;
      }

      if (this.propName !== undefined) {
        result.propName = this.propName;
      }

      if (!result.propName && this.opResult.isFindRef()) {
        result.propName = 'objectId';
      }

      return result;
    }
  }]);
  return OpResultValueReference;
}();

exports.OpResultValueReference = OpResultValueReference;
