// backendless-coderunner/lib/server-code/runners/cloud-master.js
'use strict'

const logger = require('../../util/logger')
const MessagesBroker = require('../services/messages-broker')
const WorkersBroker = require('./cloud-workers-broker')
const managementServer = require('../services/management-server')
const tasksExecutor = require('./tasks/executor')

const TASKS_CHANNEL = 'CODE_RUNNER_DRIVER'
const TASKS_CHANNEL_LP = 'JS_CR_QUEUE_LP'

const SERVICE_QUEUE_P_CR_EVENT = 'SERVICE_QUEUE_P_CR'
const CLEANUP_CODE_ALL_COMMAND = 'cleanup_code_all'

module.exports = class CloudMaster {

  constructor(options) {
    this.options = options

    this.lowPriorityThreshold = this.options.workers.lowPriorityThreshold
    this.taskRequests = {}

    this.onReadyForNextTasks = this.onReadyForNextTasks.bind(this)
    this.onTaskProcessed = this.onTaskProcessed.bind(this)
  }

  initWorkersBroker() {
    this.workersBroker = new WorkersBroker(this.options.workers, this.options)

    this.workersBroker.on(WorkersBroker.Events.READY_FOR_NEXT_TASK, this.onReadyForNextTasks)
    this.workersBroker.on(WorkersBroker.Events.TASK_PROCESSED, this.onTaskProcessed)
  }

  async startManagementServer() {
    await managementServer.start(this.options.managementHttpPort, this.workersBroker)
  }

  async startMessageBroker() {
    this.messageBroker = new MessagesBroker({
      connection        : this.options.backendless.msgBroker,
      compressionEnabled: this.options.compression.prod,
      gettersCount      : this.lowPriorityThreshold ? 2 : 1,
    })

    this.messageBroker.on('error', this.exitOnError)
    this.messageBroker.on('reconnect', this.onReadyForNextTasks)

    await this.messageBroker.init()

    this.messageBroker.subscribe(SERVICE_QUEUE_P_CR_EVENT, message => {
      if (message.command === CLEANUP_CODE_ALL_COMMAND) {
        this.workersBroker.killAllAppWorkers(message.applicationId, 'New Business Logic for app has been deployed.')
      }
    })
  }

  async start() {
    logger.info(`Starting Backendless ${this.options.label || 'Cloud'} Code Runner for JS`)
    logger.info(`Backendless Repository Path is set to [${this.options.backendless.repoPath}]`)

    this.initWorkersBroker()

    await this.startMessageBroker()
    await this.startManagementServer()

    this.onReadyForNextTasks()
  }

  onReadyForNextTasks() {
    const currentWorkersLoad = this.workersBroker.getCurrentLoad()
    const availableWorkersCount = this.workersBroker.getAvailableWorkersCount()
    const totalWorkersCount = this.workersBroker.getTotalWorkersCount()

    logger.info('Ready and waiting for Server Code tasks..', {
      load     : `${parseInt(currentWorkersLoad * 10000) / 100}%`,
      available: availableWorkersCount,
      total    : totalWorkersCount,
    })

    if (availableWorkersCount > 0) {
      this.waitAndProcessNextTask(TASKS_CHANNEL)
    }

    if (this.lowPriorityThreshold && this.lowPriorityThreshold > currentWorkersLoad && availableWorkersCount > 1) {
      /**
       * subscribe to LP Queue only if:
       *  - workers.lowPriorityThreshold is configured and value is more than zero
       *  - current workers load less than workers.lowPriorityThreshold
       *  - and there at least 2 available workers to process incoming tasks
       * **/
      this.waitAndProcessNextTask(TASKS_CHANNEL_LP)
    }
  }

  waitAndProcessNextTask(tasksChannel) {
    if (!this.taskRequests[tasksChannel]) {
      this.taskRequests[tasksChannel] = Promise.resolve()
        .then(() => this.messageBroker.getTask(tasksChannel))
        .then(async task => {
          task = Object.assign(task, { cacheable: tasksExecutor.isTaskCacheable(task) })

          const resetRequestBefore = this.workersBroker.getAvailableWorkersCount() > 2

          if (resetRequestBefore) {
            this.taskRequests[tasksChannel] = null
          }

          const worker = await this.workersBroker.getWorkerForTask(task.applicationId, task.cacheable)

          if (!resetRequestBefore) {
            this.taskRequests[tasksChannel] = null
          }

          this.workersBroker.processTask(task, worker)
        })
        .catch(() => this.taskRequests[tasksChannel] = null)
    }
  }

  async onTaskProcessed(task, result) {
    await this.messageBroker.setTaskResult(task, result)
  }

  exitOnError(error) {
    logger.error(error)

    process.exit(1)
  }

  async stop() {
    if (!this.stopped) {
      this.stopped = true

      await this.messageBroker.stopGetters()

      await this.workersBroker.stop()

      await this.messageBroker.stopSetters()
    }
  }
}
