// backendless/lib/request/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _backendlessRequest = _interopRequireDefault(require("backendless-request"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var Methods = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
  PATCH: 'PATCH'
};

var Request = /*#__PURE__*/function () {
  function Request(app) {
    (0, _classCallCheck2["default"])(this, Request);
    this.app = app;
    this.Methods = Methods;
  }

  (0, _createClass2["default"])(Request, [{
    key: "send",
    value: function send(options) {
      _backendlessRequest["default"].verbose = !!this.app.debugMode;
      _backendlessRequest["default"].XMLHttpRequest = this.app.XMLHttpRequest;
      var method = options.method.toLowerCase();
      var headers = options.headers || {};
      var userToken = this.app.getCurrentUserToken();

      if (userToken) {
        headers['user-token'] = userToken;
      }

      if (options.queryString) {
        options.url = "".concat(options.url, "?").concat(options.queryString);
      }

      return _backendlessRequest["default"][method](options.url, options.data).set(headers).query(options.query).form(options.form);
    }
  }, {
    key: "get",
    value: function get(options) {
      return this.send(_objectSpread(_objectSpread({}, options), {}, {
        method: Methods.GET
      }));
    }
  }, {
    key: "post",
    value: function post(options) {
      return this.send(_objectSpread(_objectSpread({}, options), {}, {
        method: Methods.POST
      }));
    }
  }, {
    key: "put",
    value: function put(options) {
      return this.send(_objectSpread(_objectSpread({}, options), {}, {
        method: Methods.PUT
      }));
    }
  }, {
    key: "delete",
    value: function _delete(options) {
      return this.send(_objectSpread(_objectSpread({}, options), {}, {
        method: Methods.DELETE
      }));
    }
  }]);
  return Request;
}();

exports["default"] = Request;
