// backendless/lib/logging/logger.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var Logger = /*#__PURE__*/function () {
  function Logger(name, logging) {
    (0, _classCallCheck2["default"])(this, Logger);
    this.name = name;
    this.logging = logging;
  }

  (0, _createClass2["default"])(Logger, [{
    key: "debug",
    value: function debug(message) {
      return this.logging.push(this.name, 'DEBUG', message);
    }
  }, {
    key: "info",
    value: function info(message) {
      return this.logging.push(this.name, 'INFO', message);
    }
  }, {
    key: "warn",
    value: function warn(message, exception) {
      return this.logging.push(this.name, 'WARN', message, exception);
    }
  }, {
    key: "error",
    value: function error(message, exception) {
      return this.logging.push(this.name, 'ERROR', message, exception);
    }
  }, {
    key: "fatal",
    value: function fatal(message, exception) {
      return this.logging.push(this.name, 'FATAL', message, exception);
    }
  }, {
    key: "trace",
    value: function trace(message) {
      return this.logging.push(this.name, 'TRACE', message);
    }
  }]);
  return Logger;
}();

exports["default"] = Logger;
